File: Sources/Sideros/Models/Lfm2Moe.swift
import Foundation
import MLX

public struct Lfm2MoeActivations {
    public let embeddings: MLXArray
    public let blocks: [MLXArray]
    public let norm: MLXArray
    public let logits: MLXArray
}

/// LFM2.5: a hybrid decoder where most layers are a gated short conv and every fourth
/// is GQA, and the MLP past the first two layers is a mixture of 32 experts, 4 per
/// token. The conv layers cache a sliding window instead of keys and values.
public struct Lfm2Moe: CausalLM {
    private let config: Lfm2MoeConfig
    private let parameters: [String: MLXArray]
    /// One closure per layer: the dense MLPs compiled as in Qwen, the sparse ones
    /// routing through gatherMM.
    let mlps: [(MLXArray) -> MLXArray]
    /// One compiled (x, window) → (y, window) closure per conv layer: 18 layers of
    /// tiny elementwise ops per step is launch overhead a fused trace erases.
    private let convs: [(([MLXArray]) -> [MLXArray])?]

    public init(config: Lfm2MoeConfig, parameters: [String: MLXArray]) {
        self.config = config

        // One fused projection per attention layer instead of three, as in Qwen3.
        var parameters = parameters
        for (layer, type) in config.layerTypes.enumerated() where type == "full_attention" {
            let p = "model.layers.\(layer).self_attn."
            let parts = ["q_proj.", "k_proj.", "v_proj."].map { parameters[p + $0 + "weight"]! }
            parameters[p + "qkv_proj.weight"] = concatenated(parts, axis: 0)
        }
        // Gate and up fused the same way, row-aligned along each expert's output
        // axis: one gather reads both. Materialized here (the concat is real work,
        // and a compiled closure would otherwise re-run it every token) and the
        // separate stacks dropped, or both copies would stay resident.
        for layer in config.numDenseLayers..<config.numHiddenLayers {
            let p = "model.layers.\(layer).feed_forward.experts."
            let fused = concatenated(
                [parameters[p + "w1.weight"]!, parameters[p + "w3.weight"]!], axis: 1)
            eval(fused)
            parameters[p + "w13.weight"] = fused
            parameters.removeValue(forKey: p + "w1.weight")
            parameters.removeValue(forKey: p + "w3.weight")
        }
        self.parameters = parameters

        convs = config.layerTypes.enumerated().map { layer, type in
            guard type == "conv" else { return nil }
            let p = "model.layers.\(layer).conv."
            let inProj = Linear(p + "in_proj", parameters: parameters, quantization: nil)
            let outProj = Linear(p + "out_proj", parameters: parameters, quantization: nil)
            let weight = parameters[p + "conv.weight"]!
            // The conv kernel unrolled: `convLCache` taps of [hidden].
            let taps = (0..<config.convLCache).map { weight[0..., 0, $0] }

            // The gated short conv: in_proj splits into B, C, x; a causal depthwise
            // conv runs over B·x; C gates the result. The taps are shifted elementwise
            // products accumulated in float32 like the conv kernel would, rounding
            // once — per-tap bfloat16 rounding is what would diverge from mlx-lm.
            let body: ([MLXArray]) -> [MLXArray] = { inputs in
                let (x, window) = (inputs[0], inputs[1])
                let length = x.dim(1)
                let parts = split(inProj(x), parts: 3, axis: -1)
                let bx = parts[0] * parts[2]
                let padded = concatenated([window, bx], axis: 1)
                let lifted = padded.asType(.float32)
                var conv = lifted[0..., ..<length, 0...] * taps[0]
                for j in 1..<taps.count {
                    conv = conv + lifted[0..., j..<(j + length), 0...] * taps[j]
                }
                return [
                    outProj(parts[1] * conv.asType(bx.dtype)),
                    padded[0..., length..., 0...],
                ]
            }
            // Only the one-token step compiles: its shape never changes, so it traces
            // once. Compiling prefill would re-trace per prompt length. With the
            // standard kernel of 3 the step goes through the fused conv dispatch
            // (see ConvMix.swift) and only out_proj stays a matmul.
            let step: ([MLXArray]) -> [MLXArray]
            if config.convLCache == 3 {
                let inW = parameters[p + "in_proj.weight"]!
                let flatTaps = weight.squeezed(axis: 1).reshaped(-1)
                eval(flatTaps)
                let hidden = config.hiddenSize
                step = compile { inputs in
                    let (gated, window) = convMix(
                        inputs[0].reshaped(-1), weights: inW, taps: flatTaps,
                        window: inputs[1].reshaped(2, hidden))
                    return [outProj(gated.reshaped(1, 1, hidden)), window.reshaped(1, 2, hidden)]
                }
            } else {
                step = compile(body)
            }
            return { inputs in inputs[0].dim(1) == 1 ? step(inputs) : body(inputs) }
        }

        mlps = (0..<config.numHiddenLayers).map { layer in
            let p = "model.layers.\(layer).feed_forward."
            guard layer >= config.numDenseLayers else {
                let gate = Linear(p + "w1", parameters: parameters, quantization: nil)
                let up = Linear(p + "w3", parameters: parameters, quantization: nil)
                let down = Linear(p + "w2", parameters: parameters, quantization: nil)
                // silu spelled out for the same ABBA-deadlock reason as in Qwen.
                return compile { x in
                    let gated = gate(x)
                    return down(gated * sigmoid(gated) * up(x))
                }
            }

            let router = Linear(p + "gate", parameters: parameters, quantization: nil)
            let bias = config.useExpertBias ? parameters[p + "expert_bias"]! : nil
            // The fused stack in checkpoint layout for the custom kernels; gatherMM
            // wants [experts, in, out], so it gets the swapped view.
            let gateUpW = parameters[p + "experts.w13.weight"]!
            let downW = parameters[p + "experts.w2.weight"]!
            let k = config.numExpertsPerTok
            let split = config.numExperts - k
            let hidden = config.hiddenSize
            let inner = config.moeIntermediateSize
            let normTopk = config.normTopkProb
            let scaling = config.routedScalingFactor

            func route(_ x: MLXArray) -> (chosen: MLXArray, weights: MLXArray) {
                // Sigmoid scores route; the bias joins in float32 for selection only,
                // so it shifts which experts win but never their weights.
                let scores = sigmoid(router(x))
                let selector = bias.map { scores.asType(.float32) + $0 } ?? scores
                let chosen = argPartition(selector, kth: split, axis: -1)[0..., 0..., split...]
                var weights = takeAlong(scores, chosen, axis: -1)
                if normTopk {
                    weights = weights / (weights.sum(axis: -1, keepDims: true) + 1e-6)
                }
                return (chosen, weights)
            }

            let body: (MLXArray) -> MLXArray = { x in
                let (chosen, weights) = route(x)
                let length = x.dim(-2)
                let experts: MLXArray
                if length * k >= 64 {
                    // Prefill: group the rows by expert so gatherMM streams each
                    // expert's weights once. A pure reorder — every row is still its
                    // own gemv, so the numbers cannot move.
                    let flat = chosen.reshaped(-1)
                    let order = argSort(flat)
                    let tokens = x.reshaped(length, 1, hidden)[floorDivide(order, MLXArray(Int32(k)))]
                    let sorted = flat[order]
                    let gateUp = gatherMM(
                        tokens, gateUpW.swappedAxes(-2, -1), rhsIndices: sorted,
                        sortedIndices: true)
                    let gated = gateUp[.ellipsis, ..<inner]
                    let up = gateUp[.ellipsis, inner...]
                    let routed = gatherMM(
                        gated * sigmoid(gated) * up, downW.swappedAxes(-2, -1),
                        rhsIndices: sorted, sortedIndices: true)
                    experts = routed[argSort(order)].reshaped(1, length, k, hidden)
                } else {
                    let tokens = expandedDimensions(x, axes: [-2, -3])
                    let gateUp = gatherMM(
                        tokens, gateUpW.swappedAxes(-2, -1), rhsIndices: chosen)
                    let gated = gateUp[.ellipsis, ..<inner]
                    let up = gateUp[.ellipsis, inner...]
                    let routed = gatherMM(
                        gated * sigmoid(gated) * up, downW.swappedAxes(-2, -1),
                        rhsIndices: chosen)
                    experts = squeezed(routed, axis: -2)
                }
                return (experts * expandedDimensions(weights * scaling, axis: -1)).sum(axis: -2)
            }
            // The one-token step is three custom dispatches: routing in a single
            // simdgroup, then the gemv pair — the down kernel applies silu and the
            // routing weights while it streams, so the routed MLP runs at ~700 GB/s
            // where gatherMM manages ~450 (see MoeGemv.swift).
            let routerW = parameters[p + "gate.weight"]!
            let routeBias = bias ?? zeros([config.numExperts], dtype: .float32)
            let routeScale = MLXArray([scaling])
            let step = compile { (x: MLXArray) -> MLXArray in
                let (chosen, weights) = moeRoute(
                    x.reshaped(-1), router: routerW, bias: routeBias, scale: routeScale,
                    k: k, normalized: normTopk)
                let gateUp = moeGateUp(x.reshaped(-1), weights: gateUpW, indices: chosen)
                let routed = moeDown(
                    gateUp, weights: downW, indices: chosen, routing: weights)
                return routed.sum(axis: 0).reshaped(1, 1, hidden)
            }
            return { x in x.dim(-2) == 1 ? step(x) : body(x) }
        }
    }

    /// Total by construction: loadLfm2MoeParameters rejects any checkpoint missing a name.
    private subscript(name: String) -> MLXArray { parameters[name]! }

    private func linear(_ name: String) -> Linear {
        Linear(name, parameters: parameters, quantization: nil)
    }

    public func embed(_ tokens: MLXArray, offset: Int = 0) -> MLXArray {
        let table = Embedding("model.embed_tokens", parameters: parameters, quantization: nil)
        return table(tokens)
    }

    public func makeCache() -> [KVCache] { (0..<config.numHiddenLayers).map { _ in KVCache() } }

    public func logits(_ embeddings: MLXArray, cache: [KVCache])
        -> (logits: MLXArray, cache: [KVCache])
    {
        let decoded = decode(embeddings, cache: cache)
        return (decoded.logits, decoded.cache)
    }

    public func callAsFunction(_ ids: [Int32]) -> Lfm2MoeActivations {
        let embeddings = embed(ids)
        let decoded = decode(embeddings, cache: makeCache())
        return Lfm2MoeActivations(
            embeddings: embeddings,
            blocks: decoded.blocks,
            norm: decoded.norm,
            logits: decoded.logits
        )
    }

    private func decode(_ embeddings: MLXArray, cache: [KVCache])
        -> (blocks: [MLXArray], norm: MLXArray, logits: MLXArray, cache: [KVCache])
    {
        var cache = cache
        var x = embeddings
        var blocks: [MLXArray] = []
        for layer in 0..<config.numHiddenLayers {
            (x, cache[layer]) = block(x, layer: layer, cache: cache[layer])
            blocks.append(x)
        }

        let normed = rmsNorm(x, prefix: "model.embedding_norm")
        return (blocks, normed, linear("model.embed_tokens")(normed), cache)
    }

    func block(_ x: MLXArray, layer: Int, cache: KVCache) -> (MLXArray, KVCache) {
        let prefix = "model.layers.\(layer)."
        let normed = rmsNorm(x, prefix: prefix + "operator_norm")
        let (mixed, grown) =
            config.layerTypes[layer] == "full_attention"
            ? attention(normed, prefix: prefix, cache: cache)
            : shortConv(normed, layer: layer, cache: cache)
        let residual = x + mixed
        return (
            residual + mlps[layer](rmsNorm(residual, prefix: prefix + "ffn_norm")),
            grown
        )
    }

    func attention(_ x: MLXArray, prefix: String, cache: KVCache) -> (MLXArray, KVCache) {
        let length = x.dim(1)
        let offset = cache.offset
        let p = prefix + "self_attn."

        let queryWidth = config.numAttentionHeads * config.headDim
        let keyValueWidth = config.numKeyValueHeads * config.headDim
        let fused = linear(p + "qkv_proj")(x)
        let parts = split(fused, indices: [queryWidth, queryWidth + keyValueWidth], axis: -1)

        func heads(_ part: MLXArray, _ count: Int) -> MLXArray {
            part.reshaped(1, length, count, config.headDim).transposed(0, 2, 1, 3)
        }

        // As in Qwen3: one headDim norm vector per projection, between it and the rope.
        let queries = rope(
            rmsNorm(heads(parts[0], config.numAttentionHeads), prefix: p + "q_layernorm"),
            offset: offset)
        let (keys, values, grown) = cache.appending(
            keys: rope(
                rmsNorm(heads(parts[1], config.numKeyValueHeads), prefix: p + "k_layernorm"),
                offset: offset),
            values: heads(parts[2], config.numKeyValueHeads))

        let attended = MLXFast.scaledDotProductAttention(
            queries: queries, keys: keys, values: values,
            scale: 1 / sqrt(Float(config.headDim)), mask: length == 1 ? .none : .causal)

        let projected = linear(p + "out_proj")(
            attended.transposed(0, 2, 1, 3).reshaped(1, length, queryWidth))
        return (projected, grown)
    }

    func shortConv(_ x: MLXArray, layer: Int, cache: KVCache) -> (MLXArray, KVCache) {
        let window =
            cache.window ?? zeros([1, config.convLCache - 1, config.hiddenSize], dtype: x.dtype)
        let out = convs[layer]!([x, window])
        return (out[0], cache.slid(window: out[1], by: x.dim(1)))
    }

    func rmsNorm(_ x: MLXArray, prefix: String) -> MLXArray {
        MLXFast.rmsNorm(x, weight: self[prefix + ".weight"], eps: config.normEps)
    }

    func rope(_ x: MLXArray, offset: Int = 0) -> MLXArray {
        MLXFast.RoPE(
            x, dimensions: config.headDim, traditional: false, base: config.ropeTheta,
            scale: 1, offset: offset)
    }
}

File: packages/sideros/src/sideros/models/lfm2_moe.py
"""LFM2.5-MoE: a hybrid trunk of gated short convs and GQA, with sigmoid-routed experts.

Property names are the checkpoint's. 18 of the 24 layers mix with a depthwise causal
conv of kernel 3 (cached as a 2-row window, not keys and values); the other 6 are GQA
with q/k-norm and rope theta 5e6. Past the first two dense layers the MLP is 32 experts,
4 per token, selected by `sigmoid(logits) + expert_bias` in float32 and weighted by the
bias-free score.

The conv is unrolled into `kernel` shifted taps accumulated in float32 — bit-exact with
`conv1d`, and the only form whose one-token step is a handful of elementwise ops. The
conv window cannot be rewound; the attention layers keep their KV history and trim
normally.

Config dataclass, TypedDict and loader live here for now; the integration stage moves
them into `config.py`/`checkpoint.py`.
"""

import math
from dataclasses import dataclass
from typing import NamedTuple

import mlx.core as mx
import mlx.nn as nn

from sideros.core.cache import ConvCache, KVCache
from sideros.core.kernels.conv_mix import conv_mix, conv_mix_applies
from sideros.core.kernels.moe_gemv_dense import (
    moe_dense_down,
    moe_dense_gate_up,
    moe_gemv_dense_applies,
    moe_route_sigmoid,
)
from sideros.core.layers import SORTED_GATHER_MIN, sorted_gather, split_qkv
from sideros.models.qwen3_moe import SwitchLinear

# A/B switches for the bench: set either to False to fall back to the op path, which
# stays the parity reference. The predicates read them on every call.
CONV_MIX_FUSED = True
MOE_GEMV_DENSE_FUSED = True
# The routing kernel scores from a float32 dot where the op chain rounds the gemv to T.
# In bfloat16 that flips the selection on genuine near-ties (measured: one layer of 22 on
# a synthetic prompt, the 4th and 5th experts 6e-5 apart in a selector quantized to 3e-4).
# Set False to keep the fused gemvs while the op path decides the experts.
MOE_ROUTE_FUSED = True


@dataclass(frozen=True)
class LFM2MoEConfig:
    hidden_size: int
    intermediate_size: int
    moe_intermediate_size: int
    num_hidden_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    num_experts: int
    num_experts_per_tok: int
    num_dense_layers: int
    norm_topk_prob: bool
    use_expert_bias: bool
    routed_scaling_factor: float
    norm_eps: float
    conv_bias: bool
    conv_l_cache: int
    layer_types: tuple[str, ...]
    tie_word_embeddings: bool
    rope_theta: float
    vocab_size: int

    @property
    def head_dim(self) -> int:
        """Implicit, unlike Qwen3: always hidden / heads."""
        return self.hidden_size // self.num_attention_heads


class ShortConvWeight(nn.Module):
    """The depthwise taps, `[hidden, 1, kernel]`. Never convolved as such — the kernel is
    unrolled into shifted products, so this leaf only carries the checkpoint's weight."""

    def __init__(self, hidden: int, kernel: int, bias: bool) -> None:
        super().__init__()
        self.weight = mx.zeros((hidden, 1, kernel))
        if bias:
            self.bias = mx.zeros((hidden,))


class LFM2Conv(nn.Module):
    """in_proj splits into B, C, x; a causal depthwise conv runs over B·x; C gates it."""

    def __init__(self, config: LFM2MoEConfig) -> None:
        super().__init__()
        hidden = config.hidden_size
        self.hidden = hidden
        self.kernel = config.conv_l_cache
        self.conv = ShortConvWeight(hidden, self.kernel, config.conv_bias)
        self.in_proj = nn.Linear(hidden, 3 * hidden, bias=config.conv_bias)
        self.out_proj = nn.Linear(hidden, hidden, bias=config.conv_bias)

    def fused_step_applies(self) -> bool:
        return (
            CONV_MIX_FUSED
            and type(self.in_proj) is nn.Linear
            and "bias" not in self.in_proj
            and "bias" not in self.conv
            and conv_mix_applies(self.hidden, self.kernel, has_bias=False)
        )

    def fused_step(self, x: mx.array, cache: ConvCache) -> mx.array:
        """in_proj, B·x, the three taps and C's gate in one dispatch; out_proj follows."""
        window = cache.window
        if window is None:
            window = mx.zeros((1, self.kernel - 1, self.hidden), dtype=x.dtype)
        weight = self.in_proj.weight
        assert isinstance(weight, mx.array)
        gated, slid = conv_mix(
            x.reshape(-1), weight, self.conv.weight.reshape(-1), window.reshape(2, self.hidden)
        )
        cache.window = slid[None]
        return self.out_proj(gated.reshape(1, 1, self.hidden))

    def __call__(self, x: mx.array, cache: ConvCache) -> mx.array:
        length = x.shape[1]
        # `offset` counts tokens seen, as in every other layer type: the conv keeps only a
        # window, but a layer that never advances breaks the invariant the trunk shares.
        cache.offset += length
        if length == 1 and self.fused_step_applies():
            return self.fused_step(x, cache)
        b, c, v = mx.split(self.in_proj(x), 3, axis=-1)
        bx = b * v
        window = cache.window
        if window is None:
            window = mx.zeros((1, self.kernel - 1, self.hidden), dtype=bx.dtype)
        padded = mx.concatenate([window, bx], axis=1)

        # Accumulated in float32 and rounded once, like the conv kernel: per-tap bfloat16
        # rounding is what would diverge from the reference.
        lifted = padded.astype(mx.float32)
        taps = self.conv.weight[:, 0, :]
        conv = lifted[:, :length, :] * taps[:, 0]
        for j in range(1, self.kernel):
            conv = conv + lifted[:, j : j + length, :] * taps[:, j]
        if "bias" in self.conv:
            conv = conv + self.conv.bias

        cache.window = padded[:, length:, :]
        return self.out_proj(c * conv.astype(bx.dtype))


class LFM2Attention(nn.Module):
    def __init__(self, config: LFM2MoEConfig) -> None:
        super().__init__()
        self.heads = config.num_attention_heads
        self.kv_heads = config.num_key_value_heads
        self.head_dim = config.head_dim
        self.rope_theta = config.rope_theta
        queries = self.heads * self.head_dim
        key_values = self.kv_heads * self.head_dim
        self.qkv_proj = nn.Linear(config.hidden_size, queries + 2 * key_values, bias=False)
        self.out_proj = nn.Linear(queries, config.hidden_size, bias=False)
        self.q_layernorm = nn.RMSNorm(self.head_dim, eps=config.norm_eps)
        self.k_layernorm = nn.RMSNorm(self.head_dim, eps=config.norm_eps)

    def rope(self, x: mx.array, offset: int) -> mx.array:
        return mx.fast.rope(
            x, self.head_dim, traditional=False, base=self.rope_theta, scale=1.0, offset=offset
        )

    def split_heads(self, x: mx.array) -> tuple[mx.array, mx.array, mx.array]:
        """q/k rms-normed per head between the projection and the rotation, as in Qwen3."""
        q, k, v = split_qkv(
            self.qkv_proj(x),
            heads=self.heads,
            kv_heads=self.kv_heads,
            head_dim=self.head_dim,
        )
        return self.q_layernorm(q), self.k_layernorm(k), v

    def __call__(self, x: mx.array, cache: KVCache) -> mx.array:
        length = x.shape[1]
        offset = cache.offset
        query_width = self.heads * self.head_dim
        q, k, v = self.split_heads(x)
        queries = self.rope(q, offset)
        keys, values = cache.update_and_fetch(self.rope(k, offset), v)
        attended = mx.fast.scaled_dot_product_attention(
            queries, keys, values,
            scale=1 / math.sqrt(self.head_dim),
            mask=None if length == 1 else "causal",
        )
        return self.out_proj(attended.transpose(0, 2, 1, 3).reshape(1, length, query_width))


class LFM2DenseMLP(nn.Module):
    """w1‖w3 concatenated on the output axis at load; w2 projects back."""

    def __init__(self, hidden: int, inner: int) -> None:
        super().__init__()
        self.w13 = nn.Linear(hidden, 2 * inner, bias=False)
        self.w2 = nn.Linear(inner, hidden, bias=False)
        self.inner = inner

    def __call__(self, x: mx.array) -> mx.array:
        fused = self.w13(x)
        gated = fused[..., : self.inner]
        return self.w2(gated * mx.sigmoid(gated) * fused[..., self.inner :])


class LFM2Experts(nn.Module):
    """Gate and up block-concatenated ([w1 ‖ w3] on the output axis): read by slice."""

    def __init__(self, count: int, hidden: int, inner: int) -> None:
        super().__init__()
        self.w13 = SwitchLinear(count, hidden, 2 * inner)
        self.w2 = SwitchLinear(count, inner, hidden)
        self.inner = inner

    def __call__(self, tokens: mx.array, indices: mx.array, *, sorted_indices: bool) -> mx.array:
        fused = self.w13(tokens, indices, sorted_indices=sorted_indices)
        gated = fused[..., : self.inner]
        activated = gated * mx.sigmoid(gated) * fused[..., self.inner :]
        return self.w2(activated, indices, sorted_indices=sorted_indices)


class LFM2SparseMLP(nn.Module):
    """32 experts, 4 per token, sigmoid-routed: the float32 `expert_bias` shifts which
    experts win but never their weights, which come from the bias-free score."""

    def __init__(self, config: LFM2MoEConfig) -> None:
        super().__init__()
        self.gate = nn.Linear(config.hidden_size, config.num_experts, bias=False)
        self.experts = LFM2Experts(
            config.num_experts, config.hidden_size, config.moe_intermediate_size
        )
        if config.use_expert_bias:
            self.expert_bias = mx.zeros((config.num_experts,), dtype=mx.float32)
        self.hidden = config.hidden_size
        self.k = config.num_experts_per_tok
        self.split = config.num_experts - self.k
        self.norm_topk = config.norm_topk_prob
        self.scaling = config.routed_scaling_factor

    def route(self, x: mx.array) -> tuple[mx.array, mx.array]:
        scores = mx.sigmoid(self.gate(x))
        selector = scores.astype(mx.float32) + self.expert_bias if "expert_bias" in self else scores
        chosen = mx.argpartition(selector, kth=self.split, axis=-1)[..., self.split :]
        weights = mx.take_along_axis(scores, chosen, axis=-1)
        if self.norm_topk:
            weights = weights / (weights.sum(axis=-1, keepdims=True) + 1e-6)
        return chosen, weights * self.scaling

    def fused_step_applies(self) -> bool:
        return MOE_GEMV_DENSE_FUSED and moe_gemv_dense_applies(
            self.hidden, self.experts.inner, self.experts.w13.weight.shape[0], self.k
        )

    def fused_step(self, x: mx.array) -> mx.array:
        """Routing, both gemvs, silu and the routing weight in three dispatches."""
        row = x.reshape(-1)
        gate = self.gate.weight
        assert isinstance(gate, mx.array)
        if MOE_ROUTE_FUSED:
            bias = (
                self.expert_bias
                if "expert_bias" in self
                else mx.zeros(gate.shape[0], mx.float32)
            )
            assert isinstance(bias, mx.array)
            chosen, weights = moe_route_sigmoid(
                row,
                gate,
                bias,
                mx.array(self.scaling, mx.float32),
                self.k,
                normalized=self.norm_topk,
            )
        else:
            selected, scores = self.route(x)
            chosen, weights = selected.reshape(-1).astype(mx.uint32), scores.reshape(-1)
        gate_up = moe_dense_gate_up(row, self.experts.w13.weight, chosen)
        routed = moe_dense_down(gate_up, self.experts.w2.weight, chosen, weights)
        return routed.sum(axis=0).reshape(x.shape)

    def __call__(self, x: mx.array) -> mx.array:
        if x.size == self.hidden and self.fused_step_applies():
            return self.fused_step(x)
        chosen, weights = self.route(x)
        length = x.shape[-2]
        if length * self.k >= SORTED_GATHER_MIN:

            def apply(tokens: mx.array, experts: mx.array) -> mx.array:
                return self.experts(tokens, experts, sorted_indices=True)

            routed = sorted_gather(x, chosen, k=self.k, hidden=self.hidden, apply=apply)
        else:
            tokens = mx.expand_dims(x, (-2, -3))
            routed = self.experts(tokens, chosen, sorted_indices=False).squeeze(-2)
        return (routed * mx.expand_dims(weights, -1)).sum(axis=-2)


class LFM2Block(nn.Module):
    def __init__(self, config: LFM2MoEConfig, layer: int) -> None:
        super().__init__()
        self.attends = config.layer_types[layer] == "full_attention"
        if self.attends:
            self.self_attn = LFM2Attention(config)
        else:
            self.conv = LFM2Conv(config)
        self.feed_forward: LFM2DenseMLP | LFM2SparseMLP = (
            LFM2DenseMLP(config.hidden_size, config.intermediate_size)
            if layer < config.num_dense_layers
            else LFM2SparseMLP(config)
        )
        self.operator_norm = nn.RMSNorm(config.hidden_size, eps=config.norm_eps)
        self.ffn_norm = nn.RMSNorm(config.hidden_size, eps=config.norm_eps)

    def __call__(self, x: mx.array, cache: KVCache | ConvCache) -> mx.array:
        normed = self.operator_norm(x)
        # A block has one mixer or the other; mlx.nn.Module's __getattr__ is untyped, so
        # the branch is narrowed here.
        if self.attends:
            mixer = self.self_attn
            assert isinstance(mixer, LFM2Attention) and isinstance(cache, KVCache)
            attended = x + mixer(normed, cache)
        else:
            mixer = self.conv
            assert isinstance(mixer, LFM2Conv) and isinstance(cache, ConvCache)
            attended = x + mixer(normed, cache)
        return attended + self.feed_forward(self.ffn_norm(attended))


class LFM2MoETrunk(nn.Module):
    def __init__(self, config: LFM2MoEConfig) -> None:
        super().__init__()
        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size)
        self.layers = [LFM2Block(config, layer) for layer in range(config.num_hidden_layers)]
        self.embedding_norm = nn.RMSNorm(config.hidden_size, eps=config.norm_eps)


class LFM2MoEActivations(NamedTuple):
    embeddings: mx.array
    blocks: list[mx.array]
    norm: mx.array
    logits: mx.array


class LFM2MoE(nn.Module):
    def __init__(self, config: LFM2MoEConfig) -> None:
        super().__init__()
        self.config = config
        self.model = LFM2MoETrunk(config)
        if not config.tie_word_embeddings:
            self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)

    def make_cache(self) -> list[KVCache | ConvCache]:
        return [
            KVCache() if kind == "full_attention" else ConvCache()
            for kind in self.config.layer_types
        ]

    def activations(
        self, ids: mx.array, cache: list[KVCache | ConvCache] | None = None
    ) -> LFM2MoEActivations:
        cache = cache if cache is not None else self.make_cache()
        embeddings = self.model.embed_tokens(ids)
        x = embeddings
        blocks: list[mx.array] = []
        for block, layer_cache in zip(self.model.layers, cache, strict=True):
            x = block(x, layer_cache)
            blocks.append(x)
        normed = self.model.embedding_norm(x)
        if self.config.tie_word_embeddings:
            logits = self.model.embed_tokens.as_linear(normed)
        else:
            logits = self.lm_head(normed)
        return LFM2MoEActivations(embeddings, blocks, normed, logits)

    def __call__(self, ids: mx.array, cache: list[KVCache | ConvCache] | None = None) -> mx.array:
        return self.activations(ids, cache).logits

